File: go/domain/service/order_validator.go
package service

import (
	"taller/domain/exception"
	"taller/domain/model"
)

type OrderValidator struct{}

func NewOrderValidator() *OrderValidator {
	return &OrderValidator{}
}

func (s *OrderValidator) ValidateOrder(order model.Order, availableGrams int) error {
	if order.QuantityGrams <= 0 {
		return exception.ErrInsufficientInventory
	}

	if availableGrams < order.QuantityGrams {
		return exception.ErrInsufficientInventory
	}

	return nil
}

File: typescript/src/domain/service/order-validator.ts
import { Order } from "../model/order";
import { InsufficientInventoryError } from "../exception/inventory";

export class OrderValidator {
  validateOrder(order: Order, availableGrams: number): void {
    if (order.quantityGrams <= 0) {
      throw new InsufficientInventoryError();
    }

    if (availableGrams < order.quantityGrams) {
      throw new InsufficientInventoryError();
    }
  }
}
